dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala

```diff
@@ -299,24 +308,28 @@ class MatcherApiRoute(
     )
   )
   def upsertRate: Route =
-    (path(AssetPM) & put & withAuth) { a =>
-      entity(as[Double]) { rate =>
-        if (rate <= 0) complete(RateError(error.NonPositiveAssetRate))
-        else
-          withAsset(a) { asset =>
-            complete(
-              if (asset == Waves) RateError(error.WavesImmutableRate)
-              else {
-                val assetStr = asset.toString
-                val response = rateCache.upsertRate(asset, rate) match {
-                  case None => SimpleResponse(StatusCodes.Created, s"The rate $rate for the asset $assetStr added")
-                  case Some(pv) => SimpleResponse(StatusCodes.OK, s"The rate for the asset $assetStr updated, old value = $pv, new value = $rate")
+    (path(AssetPM) & put & withAuth) { assetOrError =>
+
+      withValidAsset(assetOrError) { a =>
+        entity(as[Double]) { rate =>
+          if (rate <= 0) complete(RateError(error.NonPositiveAssetRate))
+          else
+            withAsset(a) { asset =>
```

Comment:
`withValidX` and `withX` is a frequent pattern. We need to incapsulate the logic of `withValidX` inside `withX`